src/Quizzer.API.DAL/Repository/RepositoryApi.cs

```diff
@@ -0,0 +1,82 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Quizzer.API.DAL.Entities.Interface;
+using Quizzer.API.DAL.Repository.Interfaces;
+using System.Linq.Expressions;
+
+namespace Quizzer.API.DAL.Repository {
+    public class RepositoryApiBase<TEntity> : IRepository<TEntity>, IDisposable
+        where TEntity : class, IEntity {
+
+        private readonly DbContext _dbContext;
+
+        public RepositoryApiBase(DbContext dbContext) {
+            _dbContext = dbContext;
+        }
+
+        public void Delete(Guid entityId) => DbSet().Remove(DbSet().Single(i => i.Id == entityId));
+
+        public void Dispose() {
+            _dbContext.Dispose();
+        }
+
+        public bool Exists(Guid entityId) => DbSet().Any(e => e.Id == entityId);
+
+        public IEnumerable<TEntity> GetAll(IEnumerable<string>? navigationPropertyPaths = null) {
+            IQueryable<TEntity> query = IncludeEnumeration(navigationPropertyPaths);
+
+            var entites = query.ToList();
+
+            return entites;
+        }
+
+        public IEnumerable<TEntity> GetAll() => DbSet().ToList();
+
+        public TEntity? GetById(Guid id, string navigationPropertyPath = "") {
+            return GetById(id, new string[] { navigationPropertyPath });
+        }
+
+        public TEntity? GetById(Guid id, IEnumerable<string> navigationPropertyPaths) {
+            IQueryable<TEntity> querry = IncludeEnumeration(navigationPropertyPaths);
+
+            TEntity? entity = querry.SingleOrDefault(e => e.Id == id);
+            return entity;
+        }
+
+        public IEnumerable<TEntity> GetByPredicate(Expression<Func<TEntity, bool>> predicate, string navigationPropertyPath = "") {
+            return GetByPredicate(predicate, new string[] { navigationPropertyPath });
+        }
+
+        public IEnumerable<TEntity> GetByPredicate(Expression<Func<TEntity, bool>> predicate, IEnumerable<string> navigationPropertyPaths) {
+            IQueryable<TEntity> query = IncludeEnumeration(navigationPropertyPaths);
+
+            return query.Where(predicate).ToList();
+        }
+
+        public TEntity Insert(TEntity entity) => DbSet().Add(entity).Entity;
```

Comment:
No need to call SaveChanges?

Reply:
Save changes will be called in unit of work instance. Which will be implemented above the scope of the Cook Book demonstration. The implementation of repository is more or less the same as in ICS Active Tracker with minor changes.